### PR TITLE
VW-1278: Make API and service hostnames configurable

### PR DESCRIPTION
**Pull request by thauk-copperleaf**
Branched off @sstaszkiewicz-copperleaf 's branch and PR https://github.com/copperleaftech/spring-social-github/pull/1 so it can build on Jenkins; can rebase after https://github.com/copperleaftech/spring-social-github/pull/1 is merged.

### CODE REVIEW

File: spring-social-github/src/main/java/org/springframework/social/github/api/impl/AbstractGitHubOperations.java [hunk 1]
```diff
@@ -40,9 +42,6 @@ protected void requireAuthorization() {
 	// Using String here instead of URI so I can include braces in the path. See, e.g., RepoTemplate. [WLW]
 	protected String buildUri(String path) {
 //		return URIBuilder.fromUri(API_URL_BASE + path).build();
-		return API_URL_BASE + path;
+		return "https://" + (this.apiHostname != null ? this.apiHostname : "api.github.com") + "/api/v3/" + path;
```
Comment (echen-copperleaf) [inline spring-social-github/src/main/java/org/springframework/social/github/api/impl/AbstractGitHubOperations.java]: Using `UriComponentsBuilder` might be clearer here, plus you get validation and encoding.

Like so:
```suggestion
		return UriComponentsBuilder.newInstance()
                    .scheme("https")
                    .host(this.apiHostname != null ? this.apiHostname : "api.github.com")
                    .path("/api/v3")
                    .path(path)
                    .toUriString();
```

Comment (echen-copperleaf) [reply spring-social-github/src/main/java/org/springframework/social/github/api/impl/AbstractGitHubOperations.java]: Looks like the pattern of doing simple string concat for the URL is used in multiple places, so if we're going to use `UriComponentsBuilder` here, then we should use it everywhere else too. I noticed the logic to check if the hostname is specified or null is repeated. I wonder if we can easily put that logic in one place and just return a `UriComponent` that we can add sub-paths to. The effort required to do that might be more than the benefit. The original code was doing string concat, so I'm fine with status quo. Up to you.

**Review by echen-copperleaf — APPROVED** [review]
Left some comments about repeated logic for host name selection and URL construction. I'm fine with leaving things as is.

**Comment by thauk-copperleaf** [comment]
@echen-copperleaf 

I agree with your comments and proposed changes with `UriComponentsBuilder` in principle.

I did the changes in that way to make the diff a little simpler, optimizing for future pull upstream. Looking at spring-social-github, they do not currently make use of `UriComponentsBuilder`. Do you think using that might make the changes less likely to be taken (because they might say, if we use it there, we need to use it everywhere), or would it be a neutral effect at worst?

**Comment by echen-copperleaf** [comment]
I get the feeling that the repo owners for spring-social-github have abandoned development, and the prospect of a pull upstream is probably slim. But suppose they do consider the PR, I guess it depends on who ends up reviewing it. `UriComponentsBuilder` is relatively "new" though, first introduced in Spring 3.1, which postdates this library.

If we're going to make this fork completely our own with disregard for a pull upstream, then I would make the change, otherwise it's fine as is.

**Comment by thauk-copperleaf** [comment]
@echen-copperleaf 

Yeah, I don't like the constant null checks, either. And I think we should use `UriComponentsBuilder` inside valueworks. I just always want to keep focus on the possibility on merging upstream since that removes changes from being our responsibility, especially a small one that is still so fundamental to this library working for us.

At the same time I think it bears raising an Issue with the original project to act as a ping to see if it might wake someone up: https://github.com/spring-projects/spring-social-github/issues/39

**Pull request merged**